### PR TITLE
GICv2 Distributor Config Cache

### PR DESCRIPTION
**Pull request by danshea00**
Every time an IRQ occurs the kernel reads from a GICv2 distributor configuration register to determine whether the interrupt is edge-triggered or level-triggered. If it is edge-triggered, the pending status of the interrupt must be cleared. The read from the distributor was observed to be expensive and so this pull request implements a cache for the interrupt configuration registers.

The following seL4Bench benchmarks demonstrate the performance improvement of this change using unicore MCS as measured from userspace (100 runs of IRQUser "With context switch" benchmark):

| Platform | cycles before: mean (stddev) | cycles after: mean (stddev) | diff: absolute (%) |
|---|---|---|---|
| OdroidC2 | 1132 (7) | 904 (7) | 228 (20%) |
| OdroidC4 | 1111 (7) | 889 (6) | 222 (20%) |
| TX1 | 1019 (7) | 813 (3) | 206 (20%) |

### CODE REVIEW

**Comment by Ivan-Velickovic** [comment]
It should be noted that GICv3 seems to not suffer from this issue.

File: include/arch/arm/arch/machine/gic_v2.h [hunk 1]
```diff
@@ -124,13 +125,14 @@ struct gic_cpu_iface_map {
 
 extern volatile struct gic_dist_map *const gic_dist;
 extern volatile struct gic_cpu_iface_map *const gic_cpuiface;
+extern uint32_t gic_dist_config_cache[GIC_DIST_CONFIG_SIZE];
```
Comment (Ivan-Velickovic) [inline include/arch/arm/arch/machine/gic_v2.h:128]: There should be a comment here (as well as in the commit message) as to why the change is necessary (in my opinon).

**Review by Indanz — COMMENTED** [review]
Code looks good. However, I fear you are unnecessarily optimising bad code.

I don't think the `dist_pending_clr()` call is needed for edge-triggered interrupts. Signalled interrupts are masked before acking them. See also @kent-mcleod reply at https://github.com/kent-mcleod/seL4/pull/14.

At best another edge-triggered interrupt which came before the ack will get cleared with the current code and save one unnecessary interrupt by coalescing the two edge-triggered irqs (assuming that's wanted behaviour). However, if the interrupt comes after the ack, it will still be delivered as a new interrupt. The time window in the kernel is very small, while the time window between the signal and user space calling `seL4_IRQHandler_Ack()` is much bigger, so the current code isn't very effective.

Clearing pending IRQs for edge-triggered ones would make much more sense when enabling an edge-triggered IRQ in `maskInterrupt()`, if that's what's wanted (and then your caching code makes sense too). But that won't be necessary with EOIMode = 1, as that removes the need for masking.

With the current code I think removing `is_irq_edge_triggered()` and the special code for level-triggered IRQs is the better solution. Longer-term, EOIMode = 1 should be implemented. See also #952, #1023.

**Comment by danshea00** [comment]
Thanks for the comments @Indanz! I have a few questions.

> At best another edge-triggered interrupt which came before the ack will get cleared with the current code and save one unnecessary interrupt by coalescing the two edge-triggered irqs (assuming that's wanted behaviour). However, if the interrupt comes after the ack, it will still be delivered as a new interrupt. The time window in the kernel is very small, while the time window between the signal and user space calling seL4_IRQHandler_Ack() is much bigger, so the current code isn't very effective.

I don’t see why the time window between `dist_pending_clr()` being called and the interrupt being reenabled by `seL4_IRQHandler_Ack()` is an issue. The interrupt is disabled prior to the `dist_pending_clr()` by `maskInterrupt()`. Can an interrupt become pending even if it is disabled?

> Clearing pending IRQs for edge-triggered ones would make much more sense when enabling an edge-triggered IRQ in maskInterrupt(), if that's what's wanted (and then your caching code makes sense too). 

Are you suggesting to move `dist_pending_clr()` within `seL4_IRQHandler_Ack()` (i.e. when `maskInterrupt()` reenables the interrupt)? This doesn’t make much sense to me, given that a lower priority pending interrupt would not be handled until the userspace calls `seL4_IRQHandler_Ack()` for the higher priority pending interrupt.

https://github.com/seL4/seL4/blob/61a8ccef44eeb08ecd20ad40296340c8319adee5/src/arch/arm/machine/gic_v2.c#L93-L96

Is this the “special code for level-triggered IRQs'' you are referring to?

It seems to me like for the current code, simply removing `is_irq_edge_triggered()` would make sense. The GICv2 docs indicate that this shouldn’t affect the pending status for level-triggered interrupts as the kernel never writes to `GICD_ISPENDR`.

**Comment by Indanz** [comment]
> Can an interrupt become pending even if it is disabled?

Yes. Disabling an IRQ has no effect on its state, it only stops interrupting the CPU.

> Are you suggesting to move dist_pending_clr() within seL4_IRQHandler_Ack() (i.e. when maskInterrupt() reenables the interrupt)?

_If_ special handling of edge-triggered interrupts is wanted, then yes, I propose it should be done when doing the ACK.

> This doesn’t make much sense to me, given that a lower priority pending interrupt would not be handled until the userspace calls seL4_IRQHandler_Ack() for the higher priority pending interrupt.

That's not how it works, priorities have nothing to do with it: The `dist_pending_clr()` clears the pending state of the currently active interrupt. It only has an effect if the IRQ state is active-and-pending, which happens for edge-triggered interrupts if another edge happens on the IRQ line for the currently active interrupt. See ["Figure 6 Interrupt life cycle - edge-triggered interrupts"](https://documentation-service.arm.com/static/5f1068720daa596235e7f6ef) in the GICv3 docs, which explains it better, but it also applies to GICv2.

> Is this the “special code for level-triggered IRQs'' you are referring to?

No, I meant the if-check code that uses `is_irq_edge_triggered()`.

> The GICv2 docs indicate that this shouldn’t affect the pending status for level-triggered interrupts as the kernel never writes to GICD_ISPENDR.

The main way interrupts become pending is from HW, not SW, so that's not relevant.

Level-triggered interrupts are always active and pending because the hardware still asserts the IRQ line when the interrupt becomes active. But because seL4 disables the IRQ till user-space ACKs the interrupt, it doesn't matter. For level-triggered IRQs user space will do something with the device to make the IRQ become idle again, so when it's done and it calls `seL4_IRQHandler_Ack()`, the pending will have disappeared (calling the ACK before handling the device would be a bug).

Edge-triggered interrupts behave a lot more like seL4 notifications. For those you need to call `seL4_IRQHandler_Ack()` first before handling the device, otherwise you may miss events and that would be a bug. I would even go as far as saying that you wouldn't need to call `seL4_IRQHandler_Ack()` for edge-triggered interrupts.

Common state transition for edge-triggered interrupts:

1. Edge -> IRQ pending
2. seL4 gets the `active_irq` by reading `int_ack` -> IRQ active
3. ...
4. seL4 calls `dist_pending_clr()` -> No change in IRQ state
5. seL4 Acknowledges the interrupt by writing EIO and signals user space -> IRQ idle
6. ...
7. User space calls `seL4_IRQHandler_Ack()` -> No change in IRQ state

Rare state transition for edge-triggered interrupts when another edge happens during kernel IRQ handling:

1. Edge -> IRQ pending
2. seL4 gets the `active_irq` by reading `int_ack` -> IRQ active
3. Edge -> IRQ active & pending
4. seL4 calls `dist_pending_clr()` -> IRQ active
5. seL4 Acknowledges the interrupt by writing EIO and signals user space -> IRQ idle
6. ...
7. User space calls `seL4_IRQHandler_Ack()` -> No change in IRQ state

So the only change clearing pending has is that edges happening while seL4 is in the interrupt handler are ignored. Any edged at 6 will still cause a new interrupt, even though user space was just notified an IRQ happed:

Less rare state transition for edge-triggered interrupts when another edge happens during user space IRQ handling:

1. Edge -> IRQ pending
2. seL4 gets the `active_irq` by reading `int_ack` -> IRQ active
3. (Nothing happens)
4. seL4 calls `dist_pending_clr()` -> No change in IRQ state
5. seL4 Acknowledges the interrupt by writing EIO and signals user space -> IRQ idle
6. Edge -> IRQ pending
7. User space calls `seL4_IRQHandler_Ack()` -> No change in IRQ state, but IRQ unmasked
8. seL4 gets the interrupt by reading `int_ack` -> IRQ active

Covering the race at 3, but not the bigger window of 6 doesn't make any sense. Either cover windows 6 and do something special during `seL4_IRQHandler_Ack()`, or remove all edge-triggered special casing. I propose the latter, because edge-triggered interrupts are not expected to happen often. Making all interrupt handling slightly faster and the code simpler is the better trade-off.

I'm going to open a new pull request to remove the edge-triggered special code, with the same explanation.

**Comment by danshea00** [comment]
Thanks so much for the detailed explanation, that clears up my understanding. Feel free to close this PR when you open yours.
